# KNN Regression 

Data = cars.csv Predictors = [Cylinders, Weight, Horsepower] Target = MPG

You will remove all the observations for which MPG and Horsepower is zero. Do a scaling - do standardization of all the predictors ((X-mean)/sd) [This won't make it standard normal distributed) - This is done beacuse different variables are measured in different scales.

Keep aside one random observation from the dataset as test. (DO NOT include that observation for training the algorithm)

you need to build a kNN algorithm for regression.

kNN(trainX, trainY, testX, k)

If X follows a normal distribution with mean = mu and variance=sigma, then

Z=(X - mu)/sigma will also follow a normal distribution with mean=0 and variance=1

A normal distribution with zero mean and variance 1 is called a standard normal distribution.

In [10]:
import numpy as np
import pandas as pd                        # Importing libraries 
from sklearn import preprocessing          
cars = pd.read_csv("cars.csv")            # Importing dataset
cars.shape

(406, 9)

In [11]:
cars_4 = cars[["MPG","Cylinders","Weight","Horsepower"]]    # Selecting variables under consideration

In [12]:
cars_4.head()

MPG  Cylinders  Weight  Horsepower
0  18.0          8    3504         130
1  15.0          8    3693         165
2  18.0          8    3436         150
3  16.0          8    3433         150
4  17.0          8    3449         140

### Standardizing the predictors

In [13]:
test_data = preprocessing.scale(cars_new)     # Standardizing the predictors values using the functions of sklearn
test_data = pd.DataFrame(data = test_data, columns=['MPG','Cylinders','Horsepower','Weight']) 
test_data

MPG  Cylinders  Horsepower    Weight
0   -0.698638   1.483947    0.620540  0.664133
1   -1.083498   1.483947    0.843334  1.574594
2   -0.698638   1.483947    0.540382  1.184397
3   -0.955212   1.483947    0.536845  1.184397
4   -0.826925   1.483947    0.555706  0.924265
..        ...        ...         ...       ...
387  0.455941  -0.864014   -0.221125 -0.480448
388  2.636813  -0.864014   -0.999134 -1.364896
389  1.097374  -0.864014   -0.804632 -0.532474
390  0.584228  -0.864014   -0.415627 -0.662540
391  0.969088  -0.864014   -0.303641 -0.584501

[392 rows x 4 columns]

In [14]:
cars_new = cars_4[((cars_4.Horsepower != 0) & (cars_4.MPG != 0))]  # Removing the rows having zero values
cars_new

MPG  Cylinders  Weight  Horsepower
0    18.0          8    3504         130
1    15.0          8    3693         165
2    18.0          8    3436         150
3    16.0          8    3433         150
4    17.0          8    3449         140
..    ...        ...     ...         ...
401  27.0          4    2790          86
402  44.0          4    2130          52
403  32.0          4    2295          84
404  28.0          4    2625          79
405  31.0          4    2720          82

[392 rows x 4 columns]

### After removing 0 value for the MPG and Horsepower column in dataset, now we have 392 rows and 4 columns

### -----------------------------------------------------------------------------------------------------------------------------------------------------------

### Now splitting the dataset into training data, test data and target data

In [15]:
training_data = test_data.iloc[1:, [1,2,3]]        # Training data

training_data.head()

Cylinders  Horsepower    Weight
1   1.483947    0.843334  1.574594
2   1.483947    0.540382  1.184397
3   1.483947    0.536845  1.184397
4   1.483947    0.555706  0.924265
5   1.483947    1.607198  2.433030

In [16]:
test = test_data.iloc[0:1,[1,2,3]]        # Test data
test

Cylinders  Horsepower    Weight
0   1.483947     0.62054  0.664133

In [17]:
target = cars_new.iloc[1:,[0]]           # Target data
target.head()

MPG
1  15.0
2  18.0
3  16.0
4  17.0
5  15.0

### Function for calculating Euclidean distance

### Now scale using sklearn

In [18]:
def euclidean(p,q):
    d = 0

    for i in range(len(p)):
        d = d + (p[i]-q[i])**2       # Calculating euclidean distance between 

    return(np.sqrt(d))

In [19]:
def knn(D, x, y, k):
    dist = []
    

    for i in range(0, len(D)):
        dist.append(euclidean(D.iloc[i,[0,1,2]],x.iloc[0,[0,1,2]]))    # Calculating Eucleadian distances

    d = dist.copy()      
    dist.sort()                  # Creating new list containing distance and sorting it 

                    
    index = []
    for i in range(0,k):
        indx = d.index(dist[i])    # Extracting 'K' number of values from dist list 
        index.append(indx)    
        d[indx] = -1               # Getting the index of the target value
       
        
    print(index)   
    

    pred_y = np.mean(y.iloc[index])     
    return(pred_y)

In [20]:
knn(training_data,test,target,5)

[288, 248, 283, 3, 247]


MPG    18.68
dtype: float64